storage,roachpb: fix the rule solver's diversity rule

BramGruneir opened the pull request:
This commit adds a diversity score between two localities and then uses this score in an updated diversity rule.
This also adds a nodeID to localities map within the store pool to be able to quickly retrieve a node's locality.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/10774)
<!-- Reviewable:end -->

BramGruneir committed:
storage,roachpb: fix the rule solver's diversity rule

This commit adds a diversity score between two localities and then uses this score in an updated diversity rule.
This also adds a nodeID to localities map within the store pool to be able to quickly retrieve a node's locality.

petermattis commented:
FYI, I'm probably not going to be able to review this today given other priorities.

BramGruneir commented:
No problem.  I'm going to be moving forward working on adding randomization
today.  So I'm not blocked.

On Nov 18, 2016 08:39, "Peter Mattis" <notifications@github.com> wrote:

> FYI, I'm probably not going to be able to review this today given other
> priorities.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/10774#issuecomment-261533573>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABihuX3EjdWjoEXFAqprfXsUMQjvIKFQks5q_aqAgaJpZM4K1222>
> .
>

petermattis commented:
Review status: 0 of 7 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/roachpb/metadata.go, line 244 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10774#-KX76MwID5BgDNQIUqHy:-KX76MwJyopd6Vv7EZoV:bt5ba8k) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d9b021291f6c37b2680658371b1bcba9c606261/pkg/roachpb/metadata.go#L244)):*
> ```Go
> 	length := len(l.Tiers)
> 	if len(other.Tiers) < length {
> 		length = len(other.Tiers)
> ```

This is probably mentioned somewhere else: why do you compare the prefix if the localities are different lengths instead of the suffix?

---

*[pkg/roachpb/metadata.go, line 251 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10774#-KX75xDt96vHoS3mBoLY:-KX75xDu9_fXy-IDvhgw:b-9zhkzn) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d9b021291f6c37b2680658371b1bcba9c606261/pkg/roachpb/metadata.go#L251)):*
> ```Go
> 		}
> 	}
> 	return 0
> ```

This returns 0 if the localities are `a,b,c` and `a,b` which doesn't quite mesh with the comment above. I'm not sure if that is a problem.

---

*[pkg/roachpb/metadata_test.go, line 161 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10774#-KX6qgcS3regj8RGfPvN:-KX6qgcS3regj8RGfPvO:bjjb1g6) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d9b021291f6c37b2680658371b1bcba9c606261/pkg/roachpb/metadata_test.go#L161)):*
> ```Go
> 	getFraction := func(num, den int) float64 {
> 		return float64(num) / float64(den)
> 	}
> ```

What is this about? Writing `2 / 3` (or `2.0 / 3.0`) in code should work, right?

---

*[pkg/roachpb/metadata_test.go, line 183 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10774#-KX6qo11vk5qsDuG-8bY:-KX6qo11vk5qsDuG-8bZ:b-aostsy) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d9b021291f6c37b2680658371b1bcba9c606261/pkg/roachpb/metadata_test.go#L183)):*
> ```Go
> 			right:    []string{"b"},
> 			expected: 1,
> 		},
> ```

Lot of boilerplate in these test cases. I point this out becomes it obscures the tests themselves. `left` and `right` could be comma separated strings which `generateLocality` splits. Then your test cases could become:

```
{
  "1 level, 1st different",
  "a", "b", 1,
}
```

I'm guessing @tamird will disagree, but I'm not a fan of trying to name these test cases. It places unnecessary burden when adding a new one. Leave out the name and this becomes:

```
{ "a", "b", 1},
{ "a,aa", "b,bb", 1},
{ "a,aa,aaa", "b,bb,bbb", 1},
{ "a", "a", 0},
....
```

---

*[pkg/storage/rule_solver.go, line 172 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10774#-KX6xxCYjpAzlt_R69P0:-KX6xxCYjpAzlt_R69P1:b-orompm) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d9b021291f6c37b2680658371b1bcba9c606261/pkg/storage/rule_solver.go#L172)):*
> ```Go
> 		return minScore * ruleDiversityWeight, true
> 	}
> 	return 0, true
> ```

This feels weird. We only get here if `len(state.existingNodeLocalities) == 0`, but that shouldn't be possible on a real range. 

---

*[pkg/storage/store_pool.go, line 616 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10774#-KX6w6BhtsQypm7FY4pk:-KX6w6BhtsQypm7FY4pl:b-pa2svo) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d9b021291f6c37b2680658371b1bcba9c606261/pkg/storage/store_pool.go#L616)):*
> ```Go
> // scores for faster lookups.
> func (sp *StorePool) getNodeLocalities(
> 	nodeIDs []roachpb.NodeID,
> ```

It looks like all the callers of this method besides tests have a `[]ReplicaDescriptor`. Better to pass in that slice rather than to create a new one.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10774)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 0 of 7 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/roachpb/metadata_test.go, line 183 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10774#-KX6qo11vk5qsDuG-8bY:-KX78xwPQx49eJSirG21:b4adpob) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d9b021291f6c37b2680658371b1bcba9c606261/pkg/roachpb/metadata_test.go#L183)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Lot of boilerplate in these test cases. I point this out becomes it obscures the tests themselves. `left` and `right` could be comma separated strings which `generateLocality` splits. Then your test cases could become:
> 
> ```
> {
>   "1 level, 1st different",
>   "a", "b", 1,
> }
> ```
> 
> I'm guessing @tamird will disagree, but I'm not a fan of trying to name these test cases. It places unnecessary burden when adding a new one. Leave out the name and this becomes:
> 
> ```
> { "a", "b", 1},
> { "a,aa", "b,bb", 1},
> { "a,aa,aaa", "b,bb,bbb", 1},
> { "a", "a", 0},
> ....
> ```
</details>
I do think having names is useful in general, but these names essentially repeat the code they describe, and so are useless except for the extent to which they make failures more greppable.

Note that spaces in test names are silently replaced with underscores by the test runner, and should be avoided.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10774)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 7 files at r1.
Review status: 2 of 7 files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

*[pkg/storage/store_pool.go, line 613 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10774#-KX7AMP9qSv7JYLX2_fk:-KX7AMP9qSv7JYLX2_fl:b-1egeui) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d9b021291f6c37b2680658371b1bcba9c606261/pkg/storage/store_pool.go#L613)):*
> ```Go
> 
> // getNodeLocalities returns the localities for the provided nodeIDs.
> // TODO(bram): consider storing a full list of all node to node diversity
> ```

what is this TODO?

---

*[pkg/storage/store_pool_test.go, line 539 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10774#-KX7Dae6TYERPyPzmDku:-KX7Dae6TYERPyPzmDkv:b-yw0qek) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d9b021291f6c37b2680658371b1bcba9c606261/pkg/storage/store_pool_test.go#L539)):*
> ```Go
> 		var locality roachpb.Locality
> 		if err := locality.Set(in); err != nil {
> 			t.Fatal(err)
> ```

this will produce really unfortunate errors.

---

*[pkg/storage/store_pool_test.go, line 548 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10774#-KX7DyPnRvCYpEYULhPM:-KX7DyPnRvCYpEYULhPN:b-hyvo5c) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d9b021291f6c37b2680658371b1bcba9c606261/pkg/storage/store_pool_test.go#L548)):*
> ```Go
> 			StoreID: 1,
> 			Node: roachpb.NodeDescriptor{
> 				NodeID:   1,
> ```

these ids are magic numbers that apparently defined the number of attributes in the locality; perhaps you could use strings.Repeat to make this less magical.

---

*[pkg/storage/store_pool_test.go, line 577 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10774#-KX7DrbGbRXvb3ciRWH6:-KX7DrbGbRXvb3ciRWH7:b1wnedp) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d9b021291f6c37b2680658371b1bcba9c606261/pkg/storage/store_pool_test.go#L577)):*
> ```Go
> 	sg.GossipStores(stores, t)
> 
> 	localities := sp.getNodeLocalities([]roachpb.NodeID{1, 2, 3, 1})
> ```

magic numbers?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10774)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: 2 of 7 files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

*[pkg/roachpb/metadata.go, line 244 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10774#-KX76MwID5BgDNQIUqHy:-KX7E8o7R3imf0IneVYg:bta7x0l) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d9b021291f6c37b2680658371b1bcba9c606261/pkg/roachpb/metadata.go#L244)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> This is probably mentioned somewhere else: why do you compare the prefix if the localities are different lengths instead of the suffix?
</details>
All localities are sorted from most global to most local.  
So the suffix is irrelevant if the prefixes are different.

While we recommend that all nodes have the same locality keys and length, there's nothing wrong with having different locality keys as long as the immediately next keys are all the same for each value.

So 
region:USA -> state:NY -> ...
region:USA -> state:WA -> ...
region:EUR -> country:UK -> ...
region:EUR -> country:France -> ...

Is perfectly fine, but just harder to recommend.
And this holds true at each level lower.

---

*[pkg/roachpb/metadata.go, line 251 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10774#-KX75xDt96vHoS3mBoLY:-KX7ElzPLkmibtEUIibv:babgg6v) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d9b021291f6c37b2680658371b1bcba9c606261/pkg/roachpb/metadata.go#L251)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> This returns 0 if the localities are `a,b,c` and `a,b` which doesn't quite mesh with the comment above. I'm not sure if that is a problem.
</details>
I'm not wedded to this.   We could just as easily consider empty tiers to be different instead of the same.  But what would we consider two empty tiers?

---

*[pkg/roachpb/metadata_test.go, line 161 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10774#-KX6qgcS3regj8RGfPvN:-KX7DyW-bus5yRCyJXgI:b-tr03lj) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d9b021291f6c37b2680658371b1bcba9c606261/pkg/roachpb/metadata_test.go#L161)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> What is this about? Writing `2 / 3` (or `2.0 / 3.0`) in code should work, right?
</details>
Indeed.  Don't know why I added the function.  Removed

---

*[pkg/roachpb/metadata_test.go, line 183 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10774#-KX6qo11vk5qsDuG-8bY:-KX7VTuhGcc1gbguMxcr:bmjff9y) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d9b021291f6c37b2680658371b1bcba9c606261/pkg/roachpb/metadata_test.go#L183)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> I do think having names is useful in general, but these names essentially repeat the code they describe, and so are useless except for the extent to which they make failures more greppable.
> 
> Note that spaces in test names are silently replaced with underscores by the test runner, and should be avoided.
</details>
I'll remove the names in this case, but I did find them useful while debugging.


Done.

---

*[pkg/storage/rule_solver.go, line 172 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10774#-KX6xxCYjpAzlt_R69P0:-KX7km59CMAwllsQGwa4:baqw0m4) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d9b021291f6c37b2680658371b1bcba9c606261/pkg/storage/rule_solver.go#L172)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> This feels weird. We only get here if `len(state.existingNodeLocalities) == 0`, but that shouldn't be possible on a real range. 
</details>
Sure,  I can remove the infinity check, but it feels wrong to do so.

---

*[pkg/storage/store_pool.go, line 613 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10774#-KX7AMP9qSv7JYLX2_fk:-KX7clJ4LarkEUUbjMyd:b-dek27u) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d9b021291f6c37b2680658371b1bcba9c606261/pkg/storage/store_pool.go#L613)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> what is this TODO?
</details>
It's to move the calculation of all node to node diversity scores into a stored map which can be recalculated only when a locality for a node is changed.  This would speed up diversity calculations.

---

*[pkg/storage/store_pool.go, line 616 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10774#-KX6w6BhtsQypm7FY4pk:-KX7dOB8fxoEsoXIpM5t:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d9b021291f6c37b2680658371b1bcba9c606261/pkg/storage/store_pool.go#L616)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> It looks like all the callers of this method besides tests have a `[]ReplicaDescriptor`. Better to pass in that slice rather than to create a new one.
</details>
Done.

---

*[pkg/storage/store_pool_test.go, line 539 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10774#-KX7Dae6TYERPyPzmDku:-KX7ir5a7bg7NMUJ0sND:b-hfpv0l) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d9b021291f6c37b2680658371b1bcba9c606261/pkg/storage/store_pool_test.go#L539)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this will produce really unfortunate errors.
</details>
removed

---

*[pkg/storage/store_pool_test.go, line 548 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10774#-KX7DyPnRvCYpEYULhPM:-KX7ip4madmJZTnw6uw6:b-nlkaoq) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d9b021291f6c37b2680658371b1bcba9c606261/pkg/storage/store_pool_test.go#L548)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> these ids are magic numbers that apparently defined the number of attributes in the locality; perhaps you could use strings.Repeat to make this less magical.
</details>
Made this a lot cleaner.  Can't avoid all numbers, but this should be a lot clearer.

---

*[pkg/storage/store_pool_test.go, line 577 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10774#-KX7DrbGbRXvb3ciRWH6:-KX7fmZD1Pug3fNd1obf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d9b021291f6c37b2680658371b1bcba9c606261/pkg/storage/store_pool_test.go#L577)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> magic numbers?
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10774)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 5 files at r2.
Review status: 1 of 7 files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[pkg/roachpb/metadata_test.go, line 162 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10774#-KX7uEdpCC9c7TsDUto3:-KX7uEdpCC9c7TsDUto4:b-b1x0fe) ([raw file](https://github.com/cockroachdb/cockroach/blob/6084bd47f2d8c390e43a3832ff624d53c2236d3d/pkg/roachpb/metadata_test.go#L162)):*
> ```Go
> 
> 	testCases := []struct {
> 		left     string // These will use generateLocality to produce a
> ```

yikes this is a bizarre comment style.

---

*[pkg/storage/store_pool_test.go, line 566 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10774#-KX7u5LNkWx-2mUwtEhm:-KX7u5LNkWx-2mUwtEhn:bmselio) ([raw file](https://github.com/cockroachdb/cockroach/blob/6084bd47f2d8c390e43a3832ff624d53c2236d3d/pkg/storage/store_pool_test.go#L566)):*
> ```Go
> 		{
> 			StoreID: 4,
> 			Node:    createDescWithLocality(2),
> ```

how is this passing? the test compares the length of localities to the store ID, and in this case they are not equal.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10774)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: 1 of 7 files reviewed at latest revision, 9 unresolved discussions.

---

*[pkg/roachpb/metadata_test.go, line 162 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10774#-KX7uEdpCC9c7TsDUto3:-KX7wnJrUxfx2S9DjDT9:b7anaax) ([raw file](https://github.com/cockroachdb/cockroach/blob/6084bd47f2d8c390e43a3832ff624d53c2236d3d/pkg/roachpb/metadata_test.go#L162)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> yikes this is a bizarre comment style.
</details>
Removed the comment entirely as it's pretty clear what's happening.

---

*[pkg/storage/store_pool_test.go, line 566 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10774#-KX7u5LNkWx-2mUwtEhm:-KX7wq4w5bqylzi7tXm1:b-q8s2ed) ([raw file](https://github.com/cockroachdb/cockroach/blob/6084bd47f2d8c390e43a3832ff624d53c2236d3d/pkg/storage/store_pool_test.go#L566)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> how is this passing? the test compares the length of localities to the store ID, and in this case they are not equal.
</details>
That is not what the test is doing.  The test is ensuring that the node locality is available for every store's node.  This has a node ID of 2.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10774)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
So I think this is ready to merge, PTAL.

---

Review status: 1 of 7 files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10774#-:-KXDITvAGkYKhYCeEyLL:bb7tcrn)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 1 of 7 files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[pkg/roachpb/metadata.go, line 244 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10774#-KX76MwID5BgDNQIUqHy:-KXf9up84f2Iw0WmRl3_:bjd8fep) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d9b021291f6c37b2680658371b1bcba9c606261/pkg/roachpb/metadata.go#L244)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

All localities are sorted from most global to most local.  
So the suffix is irrelevant if the prefixes are different.

While we recommend that all nodes have the same locality keys and length, there's nothing wrong with having different locality keys as long as the immediately next keys are all the same for each value.

So 
region:USA -> state:NY -> ...
region:USA -> state:WA -> ...
region:EUR -> country:UK -> ...
region:EUR -> country:France -> ...

Is perfectly fine, but just harder to recommend.
And this holds true at each level lower.

</blockquote></details>

It would be helpful to add a comment to this effect to the code.

---

*[pkg/roachpb/metadata.go, line 251 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10774#-KX75xDt96vHoS3mBoLY:-KXfA7mAlQeio4HvsBz7:b8e2xn6) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d9b021291f6c37b2680658371b1bcba9c606261/pkg/roachpb/metadata.go#L251)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

I'm not wedded to this.   We could just as easily consider empty tiers to be different instead of the same.  But what would we consider two empty tiers?
</blockquote></details>

I would have thought that two empty tiers would compare equal, but an empty tier and a non-empty tier would compare not equal. 

---

*[pkg/roachpb/metadata_test.go, line 178 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10774#-KXfB0kG5pu57PLa99mV:-KXfB0kG5pu57PLa99mW:b-nkjxnd) ([raw file](https://github.com/cockroachdb/cockroach/blob/e68f3025e147f79ae153b6bea16cf2a8f8f13e75/pkg/roachpb/metadata_test.go#L178)):*
> ```Go
> 		{"a,aa,aaa", "a,bb,bbb", 2.0 / 3.0},
> 		{"a,aa,aaa", "a,bb,aaa", 2.0 / 3.0},
> 		{"a,aa,aaa", "a,aa,bbb", 1.0 / 3.0},
> ```

You need a case for `"a,aa,aaa", "a,aa"`. Per my other comment, I think `expected` should be `1/3`.

---

*[pkg/roachpb/metadata_test.go, line 183 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10774#-KXfAr_SF4hMNbFdPa2x:-KXfAr_SF4hMNbFdPa2y:be73wsu) ([raw file](https://github.com/cockroachdb/cockroach/blob/e68f3025e147f79ae153b6bea16cf2a8f8f13e75/pkg/roachpb/metadata_test.go#L183)):*
> ```Go
> 		{"a,aa,aaa,aaaa", "a,aa,bbb,aaaa", 2.0 / 4.0},
> 		{"a,aa,aaa,aaaa", "a,aa,aaa,bbbb", 1.0 / 4.0},
> 		{"a,aa,aaa,aaaa", "b,aa", 1},
> ```

This case is the same as the one immediately below.

---

*[pkg/storage/rule_solver.go, line 172 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10774#-KX6xxCYjpAzlt_R69P0:-KXfBQ8A7uKNoDSEw4kL:b8gi6mm) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d9b021291f6c37b2680658371b1bcba9c606261/pkg/storage/rule_solver.go#L172)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Sure,  I can remove the infinity check, but it feels wrong to do so.
</blockquote></details>

Shouldn't `ruleDiversity` return `1 * ruleDiversityWeight` if there are no existing localities. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10774)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: 1 of 7 files reviewed at latest revision, 9 unresolved discussions.

---

*[pkg/roachpb/metadata.go, line 244 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10774#-KX76MwID5BgDNQIUqHy:-KXfjLNmq8A9AqmZUKXe:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d9b021291f6c37b2680658371b1bcba9c606261/pkg/roachpb/metadata.go#L244)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

It would be helpful to add a comment to this effect to the code.
</blockquote></details>

Done.

---

*[pkg/roachpb/metadata.go, line 251 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10774#-KX75xDt96vHoS3mBoLY:-KXfkS6BI0EfYo7L8v63:b-x66uuv) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d9b021291f6c37b2680658371b1bcba9c606261/pkg/roachpb/metadata.go#L251)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I would have thought that two empty tiers would compare equal, but an empty tier and a non-empty tier would compare not equal. 
</blockquote></details>

Makes sense.  I've updated the score to deal handle the case

a,b,c to  a,b = 1/3

---

*[pkg/roachpb/metadata_test.go, line 178 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10774#-KXfB0kG5pu57PLa99mV:-KXfqBRJMe0f7CxgSLdF:b-k08a89) ([raw file](https://github.com/cockroachdb/cockroach/blob/e68f3025e147f79ae153b6bea16cf2a8f8f13e75/pkg/roachpb/metadata_test.go#L178)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

You need a case for `"a,aa,aaa", "a,aa"`. Per my other comment, I think `expected` should be `1/3`.
</blockquote></details>

Done.  Also "":"a" now will be 1.  So added a few extra cases too.

---

*[pkg/roachpb/metadata_test.go, line 183 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10774#-KXfAr_SF4hMNbFdPa2x:-KXfqAmmDiuxaQsPJKGC:b-hfpv0l) ([raw file](https://github.com/cockroachdb/cockroach/blob/e68f3025e147f79ae153b6bea16cf2a8f8f13e75/pkg/roachpb/metadata_test.go#L183)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

This case is the same as the one immediately below.
</blockquote></details>

removed

---

*[pkg/storage/rule_solver.go, line 172 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10774#-KX6xxCYjpAzlt_R69P0:-KXfqrz9ZHpL90WNoxME:bljnj3p) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d9b021291f6c37b2680658371b1bcba9c606261/pkg/storage/rule_solver.go#L172)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Shouldn't `ruleDiversity` return `1 * ruleDiversityWeight` if there are no existing localities. 
</blockquote></details>

yep, done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10774)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 1 of 7 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[pkg/roachpb/metadata.go, line 241 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10774#-KXfvbtN7ieeu54FmlfA:-KXfvbtN7ieeu54FmlfB:b-lnd6r7) ([raw file](https://github.com/cockroachdb/cockroach/blob/902de9fbf08915af6ae23aad5896b0649afa058f/pkg/roachpb/metadata.go#L241)):*
> ```Go
> // their localities match). This function ignores the locality tier key names
> // and only considers differences in their values.
> // All localities are sorted from most global to most local so any localities
> ```

Can you add blank lines between paragraphs in this comment? That's the common style (except for Toby).

---

*[pkg/roachpb/metadata.go, line 251 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10774#-KXfvnhbqSyI8sQ7kACU:-KXfvnhcePTREKiBFZl9:b-qlo5k) ([raw file](https://github.com/cockroachdb/cockroach/blob/902de9fbf08915af6ae23aad5896b0649afa058f/pkg/roachpb/metadata.go#L251)):*
> ```Go
> // region:EUR -> country:UK -> ...
> // region:EUR -> country:France -> ...
> // is perfectly fine, but just harder to recommend. And this holds true at each
> ```

I'm not clear what "just harder to recommend" means. What would be easier to recommend?

---

*[pkg/storage/rule_solver.go, line 172 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10774#-KX6xxCYjpAzlt_R69P0:-KXfw90oFi2FlT5DaHnt:bmiv1nn) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d9b021291f6c37b2680658371b1bcba9c606261/pkg/storage/rule_solver.go#L172)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

yep, done
</blockquote></details>

I think this code could be simplified if you initialize `minScore = 1.0`. Then you don't have a special case if `len(state.existingNodeLocalities) == 0`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10774)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: 1 of 7 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[pkg/roachpb/metadata.go, line 241 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10774#-KXfvbtN7ieeu54FmlfA:-KXgBDskWIfJn6fUJGr8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/902de9fbf08915af6ae23aad5896b0649afa058f/pkg/roachpb/metadata.go#L241)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Can you add blank lines between paragraphs in this comment? That's the common style (except for Toby).
</blockquote></details>

Done.

---

*[pkg/roachpb/metadata.go, line 251 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10774#-KXfvnhbqSyI8sQ7kACU:-KXgB8EpoeX2o6ldobPB:bfqnvcs) ([raw file](https://github.com/cockroachdb/cockroach/blob/902de9fbf08915af6ae23aad5896b0649afa058f/pkg/roachpb/metadata.go#L251)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I'm not clear what "just harder to recommend" means. What would be easier to recommend?
</blockquote></details>

Just removed it.  It was redundant with the start of the paragraph.

---

*[pkg/storage/rule_solver.go, line 172 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10774#-KX6xxCYjpAzlt_R69P0:-KXgAklna2y15oVUX4P8:b3j0m7g) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d9b021291f6c37b2680658371b1bcba9c606261/pkg/storage/rule_solver.go#L172)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I think this code could be simplified if you initialize `minScore = 1.0`. Then you don't have a special case if `len(state.existingNodeLocalities) == 0`.
</blockquote></details>

And no more infinity checks.  Nice.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10774)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 1 of 7 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10774#-:-KXgeH0Z0QkiTzzdOi_q:bnfp4nl)*
<!-- Sent from Reviewable.io -->
